File: graphs/find_city_with_smallest_num_neighbots_threshold_distance.rb
# @param {Integer} n
# @param {Integer[][]} edges
# @param {Integer} distance_threshold
# @return {Integer}
def find_the_city(n, edges, distance_threshold)
    dist = Array.new(n) { Array.new(n, Float::INFINITY) }
    n.times { |i| dist[i][i] = 0 }
    edges.each do |u, v, w|
        dist[u][v] = dist[v][u] = w
    end
    n.times do |k|
        n.times do |i|
            n.times do |j|
                dist[i][j] = [dist[i][j], dist[i][k] + dist[k][j]].min
            end
        end
    end
    dist.each_with_index
        .min_by { |row, i| [row.count { |d| d <= distance_threshold }, -i] }
        .last
end

File: graphs/find_city_with_smallest_num_neighbots_threshold_distance.py
class Solution:
    def findTheCity(self, n: int, edges: List[List[int]], distanceThreshold: int) -> int:
        dist = [[float('inf')] * n for _ in range(n)]
        for i in range(n):
            dist[i][i] = 0

        for u, v, w in edges:
            dist[u][v] = min(dist[u][v], w)
            dist[v][u] = min(dist[v][u], w)

        for k in range(n):
            for i in range(n):
                for j in range(n):
                    if dist[i][j] > dist[i][k] + dist[k][j]:
                        dist[i][j] = dist[i][k] + dist[k][j]
        min_count = float('inf')
        city = -1
        for i in range(n):
            count = sum(1 for j in range(n) if dist[i][j] <= distanceThreshold)
            if count <= min_count:
                min_count = count
                city = i
        return city